wip: choosing online / offline stores (#305)
chalk-ai/docs: choosing-online-offline-stores.mdx

---
title: Online and Offline Stores
description: Selecting an online and offline store for your environment
---

When setting up a new environment, Chalk provides several options for choosing a storage provider
for your online and offline store. The DBMS (Database Management System) that you choose will depend on your
specific requirements, as well as your cloud deployment provider.

## Online Store Options

The Chalk online store is used for low-latency serving of real-time feature values. Although Chalk's query engine
is optimized for efficient computation of feature values by compiling Python and SQL resolvers into Rust and C++,
caching computed values in the online store can further improve performance.

### GCP Online Store Options

There are three options for online stores for customers on GCP (Google Cloud Platform):

| Store       | Description                              | Performance  | Scaling    |
|-------------|------------------------------------------|--------------|------------|
| Cloud SQL   | Postgres (relational database)           | milliseconds | vertical   |
| Memorystore | Redis (in-memory key-value NoSQL store)  | microseconds | vertical   |

> | Big Table   | Distributed wide-column NoSQL database   | milliseconds | horizontal |

Reviewer: ```suggestion
| Bigtable   | Distributed wide-column NoSQL database   | milliseconds | horizontal |
```

Generally, customers will choose the data store that they already use within their data platform. However,
we would generally recommend Memorystore for low-latency queries, Cloud SQL for general query caching and

> costs, and Big Table for huge storage needs.

Reviewer: ```suggestion
costs, and Bigtable for huge storage needs.
```

### AWS

There are three options for customers on AWS (Amazon Web Services):

| Store       | Description                    | Performance  | Scaling     |
|-------------|--------------------------------|--------------|-------------|
| RDS         | Postgres (relational database) | milliseconds | vertical    |

> | Elasticache | Redis (in-memory store)        | microseconds | vertical    |

Reviewer: ```suggestion
| ElastiCache | Redis (in-memory store)        | microseconds | vertical    |
```

| DynamoDB    | Key-Value NoSQL database       | milliseconds | auto-scales |

> We generally recommend DynamoDB for high performance and Elasticache for low-latency and greater storage capacity

Reviewer: ```suggestion
We generally recommend DynamoDB for high performance and ElastiCache for its low latency and greater storage capacity
```

`high performance` is too vague

for large rows. However, some customers may choose to use RDS depending on their existing data platform.

### Azure

There are two options for customers on Azure:

| Store                          | Description                       | Performance  | Scaling     |
|--------------------------------|-----------------------------------|--------------|-------------|
| Azure Database for PostgreSQL  | Postgres (relational database)    | milliseconds | vertical    |
| Azure Cache for Redis          | Redis (key-value in-memory store) | microseconds | horizontal  |

We generally recommend the Azure Cache for Redis for a low-latency and scalable database with storage based
pricing, but some customers may choose to use Azure Database for PostgreSQL depending on their existing data platform.

## Offline Store Options

The Chalk offline store is used for storing historical feature values. The offline store also serves
data for offline queries, which can be used for analytics, batch processing, and other use cases.
We offer four options for offline stores:

| Store                 | Storage                     |
|-----------------------|-----------------------------|

> | Google Big Query      | Columnar (BigTable-based)   |

Reviewer: ```suggestion
| Google BigQuery      | Columnar (Bigtable-based)   |
```

| Amazon Redshift       | Columnar (Parquet)          |
| Snowflake             | Columnar (Micro-partitions) |
| Databricks Delta Lake | Columnar (Parquet)          |

> For customers with GCP cloud deployments, we generally recommend Google Big Query for its performance in

Reviewer: ```suggestion
For customers with GCP cloud deployments, we generally recommend Google BigQuery for its performance with
```
